### ISSUE TITLE
cider-vs-javac: dynamically loading middleware causes problems with javac

### ISSUE THREAD

**Issue by schmir** [clojure-emacs/cider-nrepl#463]
## Expected behavior

Using lein-virgil to compile java source files should work. The java compiler should report errors for source files that contain syntactic errors.

## Actual behavior

The java compiler reports IllegalAccessErrors.

## Steps to reproduce the problem
I've setup a repository at: https://github.com/schmir/cider-vs-javac

It already contains some analysis of the problem. The commands to reproduce are:
```
 git clone https://github.com/schmir/cider-vs-javac
cd cider-vs-javac
lein repl
(doit)
```

### cider-nrepl version

0.16.0-SNAPSHOT

### Java version

1.8.0_152

### Operating system

Linux, debian 9

Here's the README from the repository mentioned above:

# cider-vs-javac

This repository is used to demonstrate a bug occuring with
cider-nrepl-0.16.0-SNAPSHOT and virgil 0.1.7

virgil recompiles java source files on the fly. If a source file contains an
error, the javac compiler reports an IllegalAccessError instead of reporting the
error in the source file if cider-nrepl is also in use.

## How to reproduce

Run `lein repl` and then call `(doit)`:

```
recompiling all files in ["/home/ralf/cider-vs-javac/java-src"]
:reloading (cider-vs-javac.core)
nREPL server started on port 39323 on host 127.0.0.1 - nrepl://127.0.0.1:39323
ADD-CLASSPATH #object[java.net.URL 0x6b635b7b file:/usr/lib/jvm/oracle-java8-jdk-amd64/src.zip]
ADD-CLASSPATH #object[java.net.URL 0x3859f725 file:/usr/lib/jvm/oracle-java8-jdk-amd64/lib/tools.jar]
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.9.0
Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

cider-vs-javac.core=> (doit)
Compiling java-src
Compiling java-src-with-errors
An exception has occurred in the compiler (1.8.0_152). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.

IllegalAccessError tried to access class com.sun.tools.javac.util.Log$2 from class com.sun.tools.javac.util.Log  com.sun.tools.javac.util.Log.getWriter (Log.java:372)
cider-vs-javac.core=> 
recompiling all files in ["/home/ralf/cider-vs-javac/java-src"]
:reloading (cider-vs-javac.core)
nREPL server started on port 39323 on host 127.0.0.1 - nrepl://127.0.0.1:39323
ADD-CLASSPATH #object[java.net.URL 0x6b635b7b file:/usr/lib/jvm/oracle-java8-jdk-amd64/src.zip]
ADD-CLASSPATH #object[java.net.URL 0x3859f725 file:/usr/lib/jvm/oracle-java8-jdk-amd64/lib/tools.jar]
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.9.0
Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

cider-vs-javac.core=> (doit)
Compiling java-src
Compiling java-src-with-errors
An exception has occurred in the compiler (1.8.0_152). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.

IllegalAccessError tried to access class com.sun.tools.javac.util.Log$2 from class com.sun.tools.javac.util.Log  com.sun.tools.javac.util.Log.getWriter (Log.java:372)
cider-vs-javac.core=> 
```

## bisect

The problem does not occur with cider-nrepl 0.15.1.

With git bisect I've found out, that the issue has been introduced in [48dc5d24b0](https://github.com/clojure-emacs/cider-nrepl/commit/48dc5d24b0b20b637ba14b459549b53e5dbf5280)

If I comment out the code to cider.nrepl.middleware.uti.java/add-classpath!, the
problem disappears. In that case tools.jar and src.zip are not being added to
the classpath.

## Workaround 

Another workaround is to force loading of `cider.nrepl.middleware.util.java`  with :injections in `project.clj`:

```
   :injections [(println "INJECT: require cider.nrepl.middleware.util.java")
                (require 'cider.nrepl.middleware.util.java)]
```

`(doit)` now works:

```
(local) ~/cider-vs-javac (git)-[master] % lein repl
INJECT: require cider.nrepl.middleware.util.java
ADD-CLASSPATH #object[java.net.URL 0x5dd903be file:/usr/lib/jvm/oracle-java8-jdk-amd64/src.zip]
ADD-CLASSPATH #object[java.net.URL 0x5a6fa56e file:/usr/lib/jvm/oracle-java8-jdk-amd64/lib/tools.jar]

recompiling all files in ["/home/ralf/cider-vs-javac/java-src"]
:reloading (cider-vs-javac.core)
nREPL server started on port 39661 on host 127.0.0.1 - nrepl://127.0.0.1:39661
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.9.0
Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

cider-vs-javac.core=> (doit)
Compiling java-src
Compiling java-src-with-errors

RuntimeException /foo/Baz.java:5: error: cannot find symbol
        xxx = 5;
        ^
  symbol:   variable xxx
  location: class foo.Baz

  sun.reflect.NativeConstructorAccessorImpl.newInstance0 (NativeConstructorAccessorImpl.java:-2)
cider-vs-javac.core=> 
```

Please note the compiler giving the right error message ("cannot find symbol xxx").

**Comment by expez** [comment]
Hi

Can you try this with `[virgil "0.1.8-alpha1"]` instead?  JDK9 support was added to virgil a few days ago.

**Comment by schmir** [comment]
That doesn't help. I'm using jdk 8.

**Comment by schmir** [comment]
If i havent't been clear enough: The problem seems to be the classpath manipulation after the compiler has been used once (lein-virgil will start the compiler before the classpath manipulation happens).

**Comment by vspinu** [comment]
Do I understand correctly the core of the issue is that `require cider.nrepl.middleware.util.java` manipulates the class path after virgil starts? Sounds like virgil's problem to me. Other tools could manipulate the class path as well. Virgil should restart the compiler after the classpath change was detected.

**Comment by schmir** [comment]
Yes, the problem is the classpath manipulation after the compiler has been used.

If I start with an empty java-src directory, the problem goes away. If I remove the classpath manipulation, the problem goes away. If I do the classpath manipulation before a file is compiled, the problem goes away.

I think this is a problem with the java compiler, not with virgil itself, which just happens to use the compiler. It looks like the java compiler itself is sensitive to classpath changes after use.

What does 'restart the compiler' mean in that context? it's not like virgil reuses an existing 'compiler object' AFAICT.

**Comment by vspinu** [comment]
> What does 'restart the compiler' 

I inferred from your description that there is some sort of compiler session in place. Either some kind of `compiler object` or more likely virgil caches state which it reuses during the re-compilation. The class path being part of that state. I am just guessing here.

**Comment by schmir** [comment]
There's no virgil cache involved. I've created the standalone branch in cider-vs-javac, and can reproduce the issue with just cider-nrepl as a dependency (and some code copied from virgil to do the compilation).

see https://github.com/schmir/cider-vs-javac/blob/standalone/src/cider_vs_javac/core.clj#L86

If the first compilation also has syntax errors, the second compilation will report the right error (demonstrated in doit2)

It looks like some class is being loaded to log syntax errors, and this fails when  adding tools.jar to the classpath.

I would say the TL;DR is: You should not add tools.jar to the classpath after the compiler has been in use.

I think it would make sense for the plugin in to add some :injection that adds tools.jar to the classpath as soon as possible.

**Comment by alexander-yakushev** [comment]
@schmir @vspinu 

Disclaimer: I've run into the same issue with CIDER and runtime javac. My setup is a bit different though:

1. I use Boot and a slightly modified version of `lein-virgil`, tailored for Boot.
2. Everything works fine for me on CIDER 0.16.0, but breaks with 0.17.0-SNAPSHOT

Otherwise, it is completely identical situation: if the compilation is triggered **before** CIDER loads, I get `java.lang.IllegalAccessError: com/sun/tools/javac/util/Log$2` on subsequent recompilations. If the first compilation happens **after** CIDER loads, everything is fine.

The previous consensus was that Virgil is to blame. That's not entirely true. Virgil calls `(javax.tools.ToolProvider/getSystemJavaCompiler)` to get an instance of the compiler. If you look into the implementation of this method ([here](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/javax/tools/ToolProvider.java#115)) you'll see that the compiler instance **isn't** cached, but the compiler class **is** cached ([here](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/javax/tools/ToolProvider.java#131)). When the compiler class is loaded for the first time, it loads `tools.jar` onto the classpath. I suppose that the second loading of `tools.jar` (which CIDER performs) somehow breaks the original cached class.

## Solution

Like @schmir described, you can force CIDER loading before the compiler runs for the first time.

The way I fixed it is I rewritten the Virgil code that gets the instance of the compiler (since I control that code anyway):

```clojure
(import 'java.net.URL 'java.net.URLClassLoader)

(defn get-java-compiler
  "Returns an instance of Java compiler. Alternative
  to `(ToolProvider/getSystemJavaCompiler)` because it doesn't cache the
  compiler class. Caching creates problems when tools.jar is added onto
  classpath AFTER the compiler was run. See
  https://github.com/clojure-emacs/cider-nrepl/issues/463."
  []
  ;; Code is rewritten from javax/tools/ToolProvider.java
  (let [file (io/file (System/getProperty "java.home"))
        file (if (.equalsIgnoreCase (.getName file) "jre")
               (.getParentFile file)
               file)
        file (io/file file "lib" "tools.jar")
        urls (into-array URL  [(.toURL (.toURI file))])
        cl (URLClassLoader/newInstance urls)
        compiler-class (Class/forName "com.sun.tools.javac.api.JavacTool" false cl)]
    (.newInstance compiler-class)))

;; Then, in the code that needs a compiler,
(ToolProvider/getSystemJavaCompiler)
;; is replaced with
(get-java-compiler)
```

I'll try to push this workaround into `lein-virgil` code.

What can be done on the CIDER side, I think, is checking whether `tools.jar` hasn't been already loaded before trying to load it ourselves. Regardless of this issue, that seems like a reasonable thing to do anyway. What do you say, @bbatsov?

**Comment by schmir** [comment]
I can confirm that this workaround fixes the issue for me.

**Comment by bbatsov** [comment]
> What can be done on the CIDER side, I think, is checking whether tools.jar hasn't been already loaded before trying to load it ourselves. Regardless of this issue, that seems like a reasonable thing to do anyway. What do you say, @bbatsov?

Yeah, that makes sense. PR welcome!

> Everything works fine for me on CIDER 0.16.0, but breaks with 0.17.0-SNAPSHOT

I just don't get what change in 0.17 could have resulted in this. The middleware loading changes happened in 0.16.

**Comment by alexander-yakushev** [comment]
This can probably be closed now. My fix was merged into `lein-virgil`, so this particular problem is resolved. To fix possible future problems I've created #515.
